include/sys/zfs_vfsops.h

```diff
@@ -132,6 +140,11 @@ struct zfsvfs {
 	uint64_t	z_hold_size;	/* znode hold array size */
 	avl_tree_t	*z_hold_trees;	/* znode hold trees */
 	kmutex_t	*z_hold_locks;	/* znode hold locks */
+	/* for controlling async zfs_unlinked_drain */
+	kmutex_t	z_drain_lock;
+	kcondvar_t	z_drain_cv;
+	drain_state_t	z_drain_state;
+	boolean_t	z_drain_continue;
```

Comment:
This all feels a bit overkill.  What do you think about simply adding a `boolean_t z_draining;` with the other booleans above, and relying on the existing `z_lock` to protect it.  It looks like you also don't really need to add a cv and could simply use `task_wait()` to determine when all the pending work has completed.

Reply:
Since we want to be able to cancel an individual (running) task for example - using `taskq_wait()` as implemented would not work since it's a pool-wide taskq.
It's possible to get away from this locking and use `taskq_wait()` if I use a dedicated per zfsvfs taskq but not sure that's a good idea?

Reply:
If you were to store the dispatched `taskqid` in the `zfsvfs` you could wait for that specific task using `taskq_wait_id()`.   Cancelling it is also possible with `taskq_cancel_id()` but setting the shutdown flag is nicer.  The only complication would be that these are both Linux specific extensions to the taskq interface.

Reply:
got it, will do. Upstream should probably use Nexenta's async drain approach anyway since they have to deals with forced unmounts